apm-lambda-extension/main_test.go

```diff
@@ -324,6 +326,30 @@ func TestAPMServerHangs(t *testing.T) {
 	assert.False(t, strings.Contains(apmServerLog.Data, string(Hangs)))
 	extension.Log.Infof("Success : test took %s", time.Since(start))
 	hangChan <- struct{}{}
+	hangChan <- struct{}{}
+}
+
+// Test what happens when the APM hangs (timeout)
+func TestAPMServerRecovery(t *testing.T) {
+	eventsChannel := make(chan MockEvent, 100)
+	lambdaServer, apmServer, apmServerLog, hangChan := initMockServers(eventsChannel)
+	defer lambdaServer.Close()
+	defer apmServer.Close()
+
+	eventsChain := []MockEvent{
+		{Type: InvokeStandard, APMServerBehavior: Hangs, ExecutionDuration: 1, Timeout: 5},
+		{Type: InvokeStandard, APMServerBehavior: TimelyResponse, ExecutionDuration: 1, Timeout: 5},
+	}
+	eventQueueGenerator(eventsChain, eventsChannel)
+	go func() {
+		time.Sleep(5 * time.Second)
+		hangChan <- struct{}{}
```

Comment:
I don't follow why we need to send twice on the hangs channel. Can you explain? Also, why is there a sleep for 5 seconds and why did you change the hangs channel to be a buffered channel of size 1?

Reply:
Following your comment, I changed the test slightly : the channel size is now unspecified. The test itself works as follows : 

- The extension receives a first event, which contains a `Hangs` signal to make the APM server hang. The forwarding request to the APM times out (after 3 seconds, as per `ELASTIC_APM_DATA_FORWARDER_TIMEOUT_SECONDS`).

- Following the request timeout, the extension goes into backoff mode. The hanging request goes back into the channel (it is now at the front of the channel, since no other element is there). During the current invocation, no other data is sent to the APM server, as the transport status is unhealthy.

- The extension receives a second event, which is supposed to be a standard, `TimelyResponse` one. The main success condition of the test is if we manage to send this event successfully to the APM server.

- In the meantime, the grace period is over, and the extension can send data again. The APM data channel now contains two chunks : a `Hangs` and a `TimelyResponse`, in that order.

- The extension now tries again to send the `Hangs` signal to the APM server. The APM server is still hanging at that point (from the previous `Hangs` signal), and does not respond immediately.

- 5 seconds after `main()` started, the `time.Sleep()` is over, and we unblock the APM server twice : once for the previous `Hangs` signal, and once to handle the `Hangs` that we sent during the second invocation due to the backoff-enqueuing mechanism.

- The transport status is now healthy, the `TimelyResponse` signal is sent to the APM server, and the test is successful.

The reasoning behind the 5 seconds is that the main thing that dimensions the test is the default value of `ELASTIC_APM_DATA_FORWARDER_TIMEOUT_SECONDS` (3 seconds). We need to unblock the APM server at some point during the second invocation, and thus not earlier than 3 seconds.

Reply:
Make `HangLock` more explicit

Reply:
Shorten forwarding timeout and put a shorter sleep time (0.5-1 sec)

Reply:
Comments addressed in c6663f7